web3/contract.py

```diff
@@ -1546,3 +1580,23 @@ def get_function_by_identifier(fns, identifier):
             'Could not find any function with matching {0}'.format(identifier)
         )
     return fns[0]
+
+
+def validate_payable(transaction, abi):
+    """Raise ValidationError if non-zero ether
+    is sent to a non payable function.
+    """
+    if 'value' in transaction:
+        if transaction['value'] != 0:
+            if "payable" in abi and not abi["payable"]:
+                raise ValidationError(
+                    "Sending non-zero ether to a contract function "
+                    "with payable=False. Please ensure that "
+                    "transaction's value is 0."
+                )
+
+
+def get_function_abi_from_contract_abi(contract_abi, function_name):
+    for abi in contract_abi:
+        if abi['name'] == function_name:
```

Comment:
A contract can have multiple functions with the same name but different argument types. 
eg. 
```solidity
contract FunctionOverloadingContract {
  function someMethod(int i) public {
     // do something
  }
  function someMethod(string s) public {
     // do something
  }
}
```
`get_function_abi_from_contract_abi ` might return the wrong function in such a case.

Have a look `find_matching_fn_abi ` to understand more. 
This code makes me think that the validation should probably moved to methods like `call_contract_function`, `transact_with_contract_function`, `build_transaction_for_function`, `estimate_gas_for_function `. I have not done a complete analysis, but just throwing an idea.